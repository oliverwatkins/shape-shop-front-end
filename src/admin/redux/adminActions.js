
export function createFetchOrdersAction(auth) {
	return {
		type: Actions.FETCH_ORDERS,
		Authorization: auth,
	};
}

export function createFetchOrdersSuccessAction(data) {
	return {
		type: Actions.FETCH_ORDERS_SUCCESS,
		data,
	};
}
export function createFetchOrdersFailAction(errorMessage: string) {
	return {
		type: Actions.FETCH_ORDERS_FAIL,
		errorMessage: errorMessage
	};
}

export const Actions = {
	FETCH_ORDERS: 'FETCH_ORDERS',
	FETCH_ORDERS_SUCCESS: 'FETCH_ORDERS_SUCCESS',
	FETCH_ORDERS_FAIL: 'FETCH_ORDERS_FAIL',
};
