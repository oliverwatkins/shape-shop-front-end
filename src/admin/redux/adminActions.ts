import {Authorization} from "../../AppState";

export function createFetchOrdersAction(auth?: Authorization) {
	return {
		type: Actions.FETCH_ORDERS,
		Authorization: auth,
	};
}

export function createFetchOrdersSuccessAction(data: any) {
	return {
		type: Actions.FETCH_ORDERS_SUCCESS,
		data,
	};
}
export function createFetchOrdersFailAction(errorMessage: string) {
	return {
		type: Actions.FETCH_ORDERS_FAIL,
		errorMessage: errorMessage
	};
}



export function createUploadImageSuccessAction() {
	return {
		type: Actions.UPLOAD_IMAGE_SUCCESS,
	};
}

export function createUploadImageFailAction(errorMessage: string) {
	return {
		type: Actions.UPLOAD_IMAGE_SUCCESS,
		errorMessage: errorMessage
	};
}

export const Actions = {
	FETCH_ORDERS: 'FETCH_ORDERS',
	FETCH_ORDERS_SUCCESS: 'FETCH_ORDERS_SUCCESS',
	FETCH_ORDERS_FAIL: 'FETCH_ORDERS_FAIL',

	UPLOAD_IMAGE: 'UPLOAD_IMAGE',
	UPLOAD_IMAGE_SUCCESS: 'UPLOAD_IMAGE_SUCCESS',
	UPLOAD_IMAGE_FAIL: 'UPLOAD_IMAGE_FAIL',
};
