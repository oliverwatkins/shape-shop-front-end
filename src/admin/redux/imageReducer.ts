export const ImageActions = {
	UPLOAD_IMAGE: 'UPLOAD_IMAGE',
	UPLOAD_IMAGE_SUCCESS: 'UPLOAD_IMAGE_SUCCESS',
	UPLOAD_IMAGE_FAIL: 'UPLOAD_IMAGE_FAIL',
};

//TODO

export function createUploadImageSuccessAction() {
	return {
		type: ImageActions.UPLOAD_IMAGE_SUCCESS,
	};
}

export function createUploadImageFailAction(errorMessage: string) {
	return {
		type: ImageActions.UPLOAD_IMAGE_SUCCESS,
		errorMessage: errorMessage
	};
}

