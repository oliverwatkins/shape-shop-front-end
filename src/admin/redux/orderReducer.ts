import {Actions} from './productActions';
import type {OrderState} from "../../AppState";
import {Notify} from "../../notify";
import {DeliveryType, OrderStateType, PaymentType} from "../../AppState";

export function reducer(state: OrderState = initialState , action: any) {

	console.info("in order reducer with action " + action.type)

	switch (action.type) {
		case Actions.UPDATE_ADDRESS:
			return {
				...state,
				addressEntity: action.value
			}
		case Actions.UPDATE_CREDIT_CARD:

			let cardType = action.value.paymentMethod.card.brand;

			let m = action.value.paymentMethod.card.exp_month;
			let y = action.value.paymentMethod.card.exp_year;

			let last4 = action.value.paymentMethod.card.last4;

			return {
				...state,
				creditCardEntity: {
					type: cardType,
					expDate: "" + m + y,
					number: "xxxx-xxxxx-xxxxx-" + last4
				}
			}
		case Actions.UPDATE_PAYMENT_TYPE:
			return {
				...state,
				paymentType: action.value
			}
		case Actions.UPDATE_DELIVERY_TYPE:
			return {
				...state,
				deliveryType: action.value
			}
		case Actions.PLACE_ORDER:
			return {
				...state,
				submittingOrder: true,
			}
		case Actions.PLACE_ORDER_SUCCESS:
			return {
				...state,
				paymentType: null,
				deliveryType: null,
				addressEntity: null,
				creditCardEntity: null,
				submittingOrder: false,
			}
		case Actions.PLACE_ORDER_ERROR:

			Notify.error("Errror placing order - " + action.errorMessage)

			return {
				...state,
				submittingOrder: false,
				orderError: action.value,
			}
		default :
			return state;
	}
}

const initialState: OrderState = {
	orderState: OrderStateType.OPEN,
	paymentType: PaymentType.cash,
	deliveryType: DeliveryType.pickup,
};
