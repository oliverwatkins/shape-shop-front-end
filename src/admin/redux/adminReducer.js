import {Actions} from './adminActions';
import type {OrderState} from "../../AppState";
import {toast} from "react-toastify";
import {Notify} from "../../notify";

export function reducer(state: OrderState = initialState , action) {
	switch (action.type) {
		case Actions.FETCH_ORDERS_SUCCESS:

			toast.success("MY SUCCESS");

			return {
				...state,
				orders: action.data,
			};
		case Actions.FETCH_ORDERS_FAIL:
			Notify.error(action.errorMsg)
			return {
				...state,
				orders: [],
				orderError: action.errorMsg,
			};
		default :
			return state;
	}
}

const initialState = {
	orders: [],
};
