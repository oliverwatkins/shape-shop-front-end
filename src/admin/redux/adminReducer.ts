import {Actions} from './adminActions';
import type {OrderState} from "../../AppState";
import {Notify} from "../../notify";
import {PaymentType} from "../../AppState";
// import {PaymentType} from "../../constants";

export function reducer(state: OrderState = initialState,
						action: { type: any; data: any; errorMessage: any; }) {

	//TODO move to orderReducer
	switch (action.type) {
		case Actions.FETCH_ORDERS_SUCCESS:
			return {
				...state,
				orders: action.data,
			};
		case Actions.FETCH_ORDERS_FAIL:
			Notify.error(action.errorMessage)
			return {
				...state,
				orders: [],
				orderError: action.errorMessage,
			};
		default :
			return state;
	}
}

const initialState = {
	orders: [],
	state: "asdf", //todo convert to enum
	paymentType: PaymentType.card,
	deliveryType: "string",
};
