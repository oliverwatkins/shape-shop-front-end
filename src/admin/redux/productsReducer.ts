import {
    Actions,
} from './productActions';
import type {ProductsState} from "../../AppState";
import {extractUniqueCategories, getCategoryProducts} from "../../util/util";
import { AnyAction } from "redux";
import {createAction} from "@reduxjs/toolkit";
import {Product} from "../../AppState";

const initialState: ProductsState = {
    categories: [],
    allProducts: [],
    productsError: "",
    updatingProduct: false,
};

export const createAddProductAction = createAction<{ product: Product } >(Actions.ADD_PRODUCT);
export const createDeleteProductAction = createAction<{ product: Product } >(Actions.DELETE_PRODUCT);
export const createUpdateProductSuccessAction = createAction<{ product: Product } >(Actions.UPDATE_PRODUCT);
export const createFetchProductsSuccessAction = createAction<{ data: any }>(Actions.FETCH_PRODUCTS);
export const createUpdateProductSelection = createAction<{ value: number, productid: string } >(Actions.UPDATE_PRODUCT_SELECTION);

//TODO this is how reducers are supposed to be written. Change other reducers to this, or
// maybe even go further with "createSlice" ? https://redux-toolkit.js.org/usage/usage-with-typescript
export function productsReducer(state: ProductsState = initialState, action: AnyAction): ProductsState {
    if (createAddProductAction.match(action)) {
        let newItems = state.allProducts;
        newItems.push(action.payload.product);
        return {
            ...state,
            allProducts: newItems
            // updatingProduct: true
        };
    }
    if (createDeleteProductAction.match(action)) {
        return {
            ...state,
            allProducts: state.allProducts.filter((elem) => {
                return elem.id !== action.payload.product.id
            })
        };
    }
    if (createUpdateProductSuccessAction.match(action)) {
        const i = state.allProducts.map((elem) => {
            if (elem.id === action.payload.product.id)
                return action.payload.product;
            return elem;
        });
        return {
            ...state,
            allProducts: i
        };
    }
    if (createFetchProductsSuccessAction.match(action)) {
        let productsAndCategories = getCategoryProducts(action.payload.data)
        return {
            ...state,
            allProducts: action.payload.data, //all products
            categoryProducts: productsAndCategories,
            categories: extractUniqueCategories(action.payload.data)
        };
    }
    if (createUpdateProductSelection.match(action)) {
        return {
            ...state,
            allProducts: state.allProducts.map((item) => {
                if (item.id === action.payload.productid) {
                    return {
                        ...item,
                        "quantity": action.payload.value
                    }
                }
                return item;
            }),
        };
    }
    return state;
}



