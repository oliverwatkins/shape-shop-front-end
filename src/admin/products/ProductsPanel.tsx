import * as React from "react";
import Typography from '@mui/material/Typography';
import {AppBar, Box, CircularProgress, Tab, Tabs, Toolbar} from "@mui/material";
import IconButton from "@mui/material/IconButton";
import MenuIcon from "@mui/icons-material/Menu";
import {useEffect, useReducer, useState} from "react";
import {useAsync} from "react-async-hook";
import {AppState, Category, Product} from "../../AppState";
import {api, extractCategories} from "../../api/api";
import {addProductAction, fetchProductsSuccessAction, updateProductSuccessAction} from "../redux/productsReducer";
import ProductPanel from "./ProductPanel";
import {useDispatch, useSelector} from "react-redux";
import Button from "@mui/material/Button";
import AddCircleOutlineIcon from "@mui/icons-material/AddCircleOutline";
import ProductDialog from "./ProductDialog";
import CategoryDialog from "./CategoryDialog";
import {Notify} from "../../notify";

export default function ProductsPanel() {


    const [createCatDialogOpen, setCreateCatDialogOpen] = React.useState(false);
    let [loading, setLoading] = useState<boolean>(false);
    let [error, setError] = useState<{ message: any }>();
    // let [loading, setLoading] = useState(false);

    const categories = useSelector((state: AppState) => state.products.categories)



    const categoryProducts = useSelector((state: AppState) => state.products.categoryProducts)
    const dispatch = useDispatch()

    function a11yProps(index: number) {
        return {
            id: `simple-tab-${index}`,
            'aria-controls': `simple-tabpanel-${index}`,
        };
    }


    // setLoading(true)
    useEffect(() => {

        // const doit = (productData: Product) => {
        setLoading(true)
        // let cs = extractCategories(productData, categories);
        // productData.categories = cs;


        api.fetchProducts().then((products: Product[]) => {
            // Notify.success("Created Product " + productData.name);
            dispatch(fetchProductsSuccessAction({data: products}));
            // dispatch(addProductAction({product: productData}))
        }).catch((error: { message: any; }) => {
            Notify.error(`onRejected function called: ${error.message}`);
            setError(error)
            // throw "this is an error"
        }).finally(() => {
            setLoading(false)
            // props.handleClose();
        });




    }, []);

    const [productTabValue, setProductTabValue] = React.useState<number>(0);
    const handleProdTab = (event: any, newValue: number) => {
        setProductTabValue(newValue);
    };

    let category = categories[productTabValue];

    return (
        <>

            {createCatDialogOpen && <CategoryDialog
                                    type={"Create"}
                                    handleClose={() => setCreateCatDialogOpen(false)}
                                     open/>}

            <Box title={"Products"}>
                <Box sx={{width: '100%'}}>
                    <Box sx={{flexGrow: 1}}>
                        <AppBar position="static">
                            <Toolbar>
                                <IconButton
                                    size="large"
                                    edge="start"
                                    color="inherit"
                                    aria-label="menu"
                                    sx={{mr: 2}}
                                >
                                    <MenuIcon/>
                                </IconButton>
                                <Typography variant="h6" component="div" sx={{flexGrow: 1}}>
                                    Products
                                </Typography>
                                <Button onClick={() => setCreateCatDialogOpen(true)}  startIcon={<AddCircleOutlineIcon/>}
                                        variant={"contained"}>
                                    Create Category</Button>
                            </Toolbar>
                        </AppBar>
                    </Box>
                    {loading && <CircularProgress color="primary"/>}
                    {error && <span>ERROR product error: {error.message}</span>}

                    <Box key={"tabs"}>
                        <Tabs value={productTabValue} onChange={handleProdTab} aria-label="asdfe">
                            {



                                categories.map((category: Category, i:number) => {
                                        // return <Tab key={category.name} label={category.name} {...a11yProps(i)} />

                                    if (!category)
                                        debugger;
                                        // alert()

                                    return <Tab key={category.name} label={category.name} {...a11yProps(i)} />

                                }


                                )
                            }
                        </Tabs>
                    </Box>
                    {categoryProducts && <ProductPanel key={productTabValue} category={category} products={categoryProducts[category.name]}/>}
                </Box>
            </Box>
        </>
    )
}
