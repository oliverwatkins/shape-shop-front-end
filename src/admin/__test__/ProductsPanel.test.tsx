import React from 'react';

import {fireEvent, getByText, render, screen, within} from '@testing-library/react'
import '@testing-library/jest-dom'
import {combineReducers, createStore} from "redux";
import {productsReducer as products} from "../../admin/redux/productsReducer";
import {reducer as loginReducer} from "../../login/redux/loginReducer";

import {reducer as order} from "../../admin/redux/orderReducer";
import {reducer as admin} from "../../admin/redux/adminReducer";
import { Provider } from 'react-redux';
import {MemoryRouter} from "react-router-dom";
import ProductsPanel from "../products/ProductsPanel";

export function createTestStore() {

	return createStore(
		combineReducers({
			// @ts-ignore POS ts comnpiler
			login: loginReducer,
			products: products,
			order: order,
			admin: admin
		}),
		{
			login: {
				loginToken: {
					token: "atoken",
					role: "arole",
					username: "auser"
				},
				loggingIn: false
			},
		}
	);
}


// login: {
// 	loginToken: {
// 		token: "atoken",
// 			role: "arole",
// 			username: "auser"
// 	},
// 	loggingIn: false
// },

let productlist = [
	{
		"id": 1,
		"name": "first",
		"price": 2.1,
		"description": "my desc",
		"imageFilename": "img.jpg",
		"categories": [
			{
				"id": 0,
				"name": "catone"
			}
		]
	},
	{
		"id": 2,
		"name": "second",
		"price": 3.1,
		"description": "my desc 2",
		"imageFilename": "img2.jpg",
		"categories": [
			{
				"id": 1,
				"name": "cattwo"
			}
		]
	}
]

// jest.setTimeout(1000000)
describe('Products test', () => {

	beforeAll(() => jest.spyOn(window, 'fetch'))

	it("renders error", async () => {

		// @ts-ignore
		window.fetch.mockResolvedValueOnce({
			ok: false,
			json: async () => ([]),
		})

		// @ts-ignore
		const {container} = render(<Provider store={createTestStore()}>
			<MemoryRouter>
				<ProductsPanel/>
			</MemoryRouter>
		</Provider>);

		await screen.findByText('ERROR product error: Network response was not ok', { collapseWhitespace: true })
	});

	it("renders products correctly", async () => {

		// @ts-ignore
		window.fetch.mockResolvedValueOnce({
			ok: true,
			json: async () => (productlist),
		})

		// @ts-ignore
		const {container} = render(<Provider store={createTestStore()}>
			<MemoryRouter>
				<ProductsPanel/>
			</MemoryRouter>
		</Provider>);

		await screen.findAllByRole('heading')
		screen.debug(container);

		expect(screen.getByRole('heading')).toHaveTextContent('Category : catone')

		expect(screen.getByText('Create Category')).toBeTruthy();
		expect(screen.getByText('Add Product')).toBeTruthy();
		expect(screen.getByText('Edit Category')).toBeTruthy();
		expect(screen.getByText('Delete Category')).toBeTruthy();

		const boxes = container.getElementsByClassName('MuiCardHeader-root');
		expect(boxes.length).toBe(1)
		expect(boxes[0]).toHaveTextContent('first');

		let b = screen.getByRole('tab', {
			name: "cattwo"
		})
		fireEvent.click(b);


		const boxes2 = container.getElementsByClassName('MuiCardHeader-root');
		expect(boxes2.length).toBe(1)
		expect(boxes2[0]).toHaveTextContent('second');
	});

	it('matches snapshot', () => {
		expect(render(<Provider store={createTestStore()}>
			<MemoryRouter>
				<ProductsPanel/>
			</MemoryRouter>
		</Provider>)).toMatchSnapshot();
	});
});
