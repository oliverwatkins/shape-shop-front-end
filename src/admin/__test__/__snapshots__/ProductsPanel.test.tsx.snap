// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products test matches snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
    title="Products"
  >
    <div
      class="MuiBox-root css-8atqhb"
    >
      <div
        class="MuiBox-root css-i9gxme"
      >
        <header
          class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
          >
            <button
              aria-label="menu"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-ancrnh-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="MenuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <div
              class="MuiTypography-root MuiTypography-h6 css-1juivf6-MuiTypography-root"
            >
              Products
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="AddCircleOutlineIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                  />
                </svg>
              </span>
              Create Category
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </header>
      </div>
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiTabs-root css-1ujnqem-MuiTabs-root"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
            style="overflow: hidden; margin-bottom: 0px;"
          >
            <div
              aria-label="asdfe"
              class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
              role="tablist"
            >
              <button
                aria-controls="simple-tabpanel-0"
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                id="simple-tab-0"
                role="tab"
                tabindex="0"
                type="button"
              >
                main
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="simple-tabpanel-1"
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                id="simple-tab-1"
                role="tab"
                tabindex="-1"
                type="button"
              >
                drinks
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-2tay6y"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-1s4ev7b-MuiTypography-root"
        >
          products of category 
          main
        </h6>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-10n6lcx-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AddCircleOutlineIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
          </span>
          Add Product
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-10n6lcx-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="EditIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
            </svg>
          </span>
          Edit Category
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-hmvwos-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Delete Category
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-lg-2 css-1ljl8iy-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root admin-item-box css-bhp9pd-MuiPaper-root-MuiCard-root"
            data-testid="custom-element"
            role="fuck"
          >
            <div
              class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
            >
              <div
                class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
              >
                <span
                  class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
                >
                  Lachs-Lasagne
                </span>
              </div>
              <div
                class="MuiCardHeader-action css-sgoict-MuiCardHeader-action"
              >
                <button
                  aria-label="settings"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="MoreVertIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
            <img
              alt="lachs_spinat.jpg"
              class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
              height="250px"
              src="http://localhost:8081/images/alpenhof/lachs_spinat.jpg"
              width="300px"
            />
            <div
              class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
            >
              <button
                aria-expanded="false"
                aria-label="show more"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-3ogxn6-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ExpandMoreIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-lg-2 css-1ljl8iy-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root admin-item-box css-bhp9pd-MuiPaper-root-MuiCard-root"
            data-testid="custom-element"
            role="fuck"
          >
            <div
              class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
            >
              <div
                class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
              >
                <span
                  class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
                >
                  Calamari
                </span>
              </div>
              <div
                class="MuiCardHeader-action css-sgoict-MuiCardHeader-action"
              >
                <button
                  aria-label="settings"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="MoreVertIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
            <img
              alt="calamari.jpg"
              class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
              height="250px"
              src="http://localhost:8081/images/alpenhof/calamari.jpg"
              width="300px"
            />
            <div
              class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
            >
              <button
                aria-expanded="false"
                aria-label="show more"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-3ogxn6-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ExpandMoreIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
